File: Homework/1 & 2/Solutions/Problem (3)/problem_3.m
% This is a function to implement the Gram-Shmidt orthogonalization (GSO)
% for two arbitrary 3-Dimensional COLUMN VECTORs
% Input :
%       a => 1st arbitrary column vector (3x1)
%       b => 2nd arbitrary column vector (3x1)
% Output:
%       v1,v2 => two orthogonal column vectors
%
% To test the correctness of answer you can check 
%                   (v1.v2) = 0     
%
% Written By- SHOWREN DATTA
% Last Modified On- September 11th, 2020
%**************************************************************************

function [v1,v2] = problem_3(a,b)

% Setting v1 equal to a
v1 = a;

% Calculating (a.b)/|a|^2, which is a scalar quantity
scalar = a' * b / ( a' * a );

% Finding the direction vector 'vec' along v2 
vec = b - scalar * a;

% Finding the value of c, that ensures |v2|=|b|
c = sqrt (b' * b)/ sqrt (vec' * vec);

% Finally calculating v2 
v2 = c * vec;

end

File: Homework/1 & 2/Solutions/Problem (3)/problem_3.py
# -*- coding: utf-8 -*-
"""
Created on Tue Sep 15 12:32:03 2020

@author: sdatta8
"""
import numpy as np

def problem_3(a,b):
    a = np.array(a)
    b = np.array(b)
    # Setting v1 equal to a
    v1 = a;
    
    # Calculating (a.b)/|a|^2, which is a scalar quantity
    scalar = a.T @ b / ( a.T @ a );
    
    # Finding the direction vector 'vec' along v2 
    vec = b - scalar * a;
    
    # Finding the value of c, that ensures |v2|=|b|
    c = np.sqrt (b.T @ b)/ np.sqrt (vec.T @ vec);
    
    # Finally calculating v2 
    v2 = c * vec;
    
    return v1, v2
